app/helpers/title_helper.rb

```diff
@@ -0,0 +1,266 @@
+      caption_plain_param(query, key, val, truncate)
+      concat(tag.span(", ")) if idx < len - 1
+    end
+  end
+
+  def caption_plain_param(query, key, val, truncate)
+    label = :"query_#{key}".l
+    # Just print the label for booleans (no `true`)
+    if val == true
+      concat(tag.span(label))
+    else
+      concat(tag.span("#{label}: ")) unless CAPTION_IGNORE_KEYS.include?(key)
+      val = lookup_text_val(query, key, val, truncate)
+      concat(tag.b(val))
+    end
+  end
+
+  # The following params store IDs, but the captions are more legible
+  # if they print a relevant "name" or "title" of the record.
+  # This indexes which Lookup class to use to get the record:
+  PARAM_LOOKUPS = {
+    external_sites: :ExternalSites,
+    field_slips: :FieldSlips,
+    herbaria: :Herbaria,
+    locations: :Locations,
+    names: :Names,
+    lookup: :Names,
+    clade: :Names,
+    projects: :Projects,
+    project_lists: :ProjectSpeciesLists,
+    species_lists: :SpeciesLists,
+    by_users: :Users,
+    for_user: :Users,
+    by_author: :Users,
+    by_editor: :Users,
+    collectors: :Users,
+    members: :Users
+  }.freeze
+  # The captions with these sub-params make more sense without the keys:
+  CAPTION_IGNORE_KEYS = [:lookup, :id, :type].freeze
+  # Max number of values to display if truncated:
+  CAPTION_TRUNCATE = 3
+
+  # Tries to get a proper name for the comment target.
+  def lookup_comment_target_val(hash)
+    type, id = hash.values_at(:type, :id)
+    return unless type && id
+
+    lookup = "Lookup::#{type.pluralize}".constantize
+    lookup.new(id).titles.first
+  end
+
+  # NOTE: Can respond to special methods for certain keys.
+  # Defaults to using the lookup method defined in CAPTIONABLE_QUERY_PARAMS
+  def lookup_text_val(query, key, val, truncate)
+    return param_val_itself(val, truncate) unless PARAM_LOOKUPS.key?(key)
+
+    # Allow overrides (second param `true` means check for private methods)
+    if [:names, :lookup].include?(key)
+      tag.i(lookup_strings(query, key, truncate))
+    else
+      lookup_strings(query, key, truncate)
+    end
+  end
+
+  # For values that aren't ids
+  def param_val_itself(val, truncate)
+    if val.is_a?(Array)
+      val = val.first(CAPTION_TRUNCATE) if truncate
+      val = val.join(", ")
+    end
+    val
+  end
+
+  # The max number of named items is hardcoded here to 3.
+  def lookup_strings(query, param, truncate)
+    ids = query.params.deep_find(param)
+    lookups = if truncate
+                ids.first(CAPTION_TRUNCATE)
+              else
+                ids
+              end
+    lookup_class = "Lookup::#{PARAM_LOOKUPS[param]}".constantize
+    joined_vals = lookup_class.new(lookups).titles.join(", ")
+    return joined_vals unless truncate
+
+    truncate_joined_string(joined_vals, ids)
+  end
+
+  def truncate_joined_string(joined_vals, ids)
+    if joined_vals.length > 100
+      joined_vals = "#{joined_vals[0...97]}..."
```

Comment:
Not covered.

Reply:
Ok, yes. Will need to send a long query

Reply:
Just want to say that writing this test improved the helper. I had forgotten to add the ellipsis after truncated text vals (looked-up text results already get it) and writing the expectation made me notice.